Applications/Spire/Source/Ui/ScrollableListBox.cpp

```diff
@@ -0,0 +1,252 @@
+#include "Spire/Ui/ScrollableListBox.hpp"
+#include <QCoreApplication>
+#include <QResizeEvent>
+#include <QHBoxLayout>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Spire/ValueModel.hpp"
+#include "Spire/Ui/ListView.hpp"
+#include "Spire/Ui/ScrollBar.hpp"
+
+using namespace Spire;
+using namespace Spire::Styles;
+
+ScrollableListBox::ScrollableListBox(ListView* list_view, QWidget* parent)
+    : ScrollBox(make_body(), parent)
+    , m_list_view(list_view) {
+  m_list_view->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  m_list_view->set_edge_navigation(ListView::EdgeNavigation::CONTAIN);
+  auto layout = [=] () -> QBoxLayout* {
+    if(is_horizontal_layout()) {
+      return new QHBoxLayout(m_body);
+    }
+    return new QVBoxLayout(m_body);
+  }();
+  layout->setSizeConstraint(QLayout::SetFixedSize);
+  layout->setContentsMargins({});
+  layout->setSpacing(0);
+  layout->addWidget(list_view);
+  if(is_horizontal_layout()) {
+    m_padding_size = get_vertical_scroll_bar().sizeHint().width();
+  } else {
+    m_padding_size = get_horizontal_scroll_bar().sizeHint().height();
+  }
+  layout->addSpacing(m_padding_size);
+  m_scroll_bar_padding = layout->itemAt(1);
+  if(is_horizontal_layout()) {
+    set_horizontal(ScrollBox::DisplayPolicy::NEVER);
+    set_vertical(ScrollBox::DisplayPolicy::ON_OVERFLOW);
+  } else {
+    set_horizontal(ScrollBox::DisplayPolicy::ON_OVERFLOW);
+    set_vertical(ScrollBox::DisplayPolicy::NEVER);
+  }
+  auto style = get_style(*this);
+  style.get(Any()).
+    set(BackgroundColor(QColor::fromRgb(0xFF, 0xFF, 0xFF))).
+    set(border(scale_width(1), QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+  set_style(*this, std::move(style));
+  m_list_view_current_connection =
+    m_list_view->get_current_model()->connect_current_signal(
+    [=] (const auto& current) { on_current(current); });
+}
+
+QSize ScrollableListBox::sizeHint() const {
+  auto border_size = get_border_size();
+  if(is_horizontal_layout()) {
+    return m_list_view->sizeHint() + QSize(m_padding_size, 0) + border_size;
+  }
+  return m_list_view->sizeHint() + QSize(0, m_padding_size) + border_size;
+}
+
+void ScrollableListBox::keyPressEvent(QKeyEvent* event) {
+  switch(event->key()) {
+    case Qt::Key_Home:
+    case Qt::Key_End:
+    case Qt::Key_PageUp:
+    case Qt::Key_PageDown:
+    case Qt::Key_Down:
+    case Qt::Key_Left:
+    case Qt::Key_Right:
+    case Qt::Key_Up:
+      return;
+  }
+  ScrollBox::keyPressEvent(event);
+}
+
+void ScrollableListBox::resizeEvent(QResizeEvent* event) {
+  auto border_size = get_border_size();
+  auto update_list_view_width = [=] (int width) {
```

Comment:
Can some of these lambdas (and some others in the update_ranges) function be put into private methods?

Reply:
They are not used by other functions, so I think it's reasonable that they are defined in this function.